database.go

```diff
@@ -166,3 +164,71 @@ func (e ErrRequiredAttributeMissing) Error() string {
 		string(e),
 	)
 }
+
+type BinariesUsages map[int][]*BinaryReference
+
+func (db *Database) getBinaries() *Binaries {
+	if db.Header.IsKdbx4() {
+		return &db.Content.InnerHeader.Binaries
+	}
+
+	return &db.Content.Meta.Binaries
+}
+
+func (db *Database) cleanupBinaries() int {
+	usages := db.getBinariesUsages()
+	binaries := *db.getBinaries()
+	updated := Binaries{}
+	removed := 0
+	counter := 0
+
+	for _, binary := range binaries {
+		if _, ok := usages[binary.ID]; ok {
+			for _, ref := range usages[binary.ID] {
+				ref.Value.ID = counter
+			}
+			binary.ID = counter
+			updated = append(updated, binary)
+			counter++
+		} else {
+			removed++
+		}
+	}
+
+	*db.getBinaries() = append(Binaries{}, updated...)
+	return removed
+}
+
+func (db *Database) getBinariesUsages() BinariesUsages {
+	result := BinariesUsages{}
+
+	for _, ref := range db.getAllBinariesReferenes(&db.Content.Root.Groups[0]) {
+		result[ref.Value.ID] = append(result[ref.Value.ID], ref)
+	}
+
+	return result
+}
+
+func (db *Database) getAllBinariesReferenes(parent *Group) []*BinaryReference {
+	result := []*BinaryReference{}
+
+	if parent != nil {
+		for _, entry := range parent.Entries {
+			for i := range entry.Binaries {
+				result = append(result, &entry.Binaries[i])
+			}
+			for _, history := range entry.Histories {
+				for _, historyEntry := range history.Entries {
+					for i := range historyEntry.Binaries {
+						result = append(result, &historyEntry.Binaries[i])
+					}
+				}
+			}
+		}
```

Comment:
Don't know how to rid of this nested loops.

Reply:
By the way is it correct that entry contains collection of histories, every of which is collection of entries? I as user expect single history with collection of entries, but I didn't investigated kdbx format so deep...

Reply:
Each entry contains its own history. There is no one all encompasing history, because you'd have to add a reference back to the latest value from the history.
That works well if you want to scroll through a list of changes you made. It works really badly if you actually want to e.g. revert to the previous version.
Also the general idea when changing an entry would be:
On the first change, move the entry into the history and make a change to the current one. After it was persisted next the same would happen again. Therefore if you were to remove a file, it should live on until you cleared the history. `gokeepasslib` does not provide convenience functions for it, because the goal of `gokeepasslib` is not to be a keepass client, but rather a library that allows reading and writing of KDBX files. What you make of it would be up to you. - At least that used to be the idea.

File attachments required more optimization during storage, which is why they are not directly added to the files. Otherwise the general idea (based on my understanding of the file format) is that an Entry is responsible for its attachments. That thought in mind also made it seem somewhat weird to me when you mentioned that you want to have a list of all files that you can scroll through.

The unit of relevance is an entry. The entry has a version, the entry has all kinds of other data attached. Entries are organized in groups to make it slightly more searchable and allow reasonably well to re-use entry titles